File: tests/Web.Tests.Playwright/tests/pages/home.page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base.page';

/**
 * Home Page Object Model
 */
export class HomePage extends BasePage {
  readonly pageHeading: Locator;
  readonly welcomeMessage: Locator;
  readonly recentArticles: Locator;
  readonly featuredContent: Locator;

  constructor(page: Page) {
    super(page);
    this.pageHeading = page.locator('h1, h2').first();
    this.welcomeMessage = page.locator('text=/welcome/i').first();
    this.recentArticles = page.locator('[data-testid="recent-articles"], .recent-articles');
    this.featuredContent = page.locator('[data-testid="featured-content"], .featured-content');
  }

  /**
   * Navigate to home page
   */
  async goto(): Promise<void> {
    await super.goto('/');
  }

  /**
   * Check if welcome message is visible
   */
  async hasWelcomeMessage(): Promise<boolean> {
    try {
      return await this.welcomeMessage.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Get the main heading text
   */
  async getHeadingText(): Promise<string> {
    return await this.pageHeading.textContent() || '';
  }

  /**
   * Check if recent articles section is visible
   */
  async hasRecentArticles(): Promise<boolean> {
    try {
      return await this.recentArticles.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }
}

File: tests/Web.Tests.Playwright/PageObjects/HomePage.cs
namespace Web.Tests.Playwright.PageObjects;

/// <summary>
/// Home Page Object Model
/// </summary>
public class HomePage : BasePage
{
    private readonly ILocator _pageHeading;
    private readonly ILocator _welcomeMessage;
    private readonly ILocator _recentArticles;
    private readonly ILocator _featuredContent;

    public HomePage(IPage page) : base(page)
    {
        _pageHeading = page.Locator("h1, h2").First;
        _welcomeMessage = page.Locator("text=/welcome/i").First;
        _recentArticles = page.Locator("[data-testid='recent-articles'], .recent-articles");
        _featuredContent = page.Locator("[data-testid='featured-content'], .featured-content");
    }

    /// <summary>
    /// Navigate to home page
    /// </summary>
    public override async Task GotoAsync(string path = "/")
    {
        await base.GotoAsync("/");
    }

    /// <summary>
    /// Check if a welcome message is visible
    /// </summary>
    public async Task<bool> HasWelcomeMessageAsync()
    {
        try
        {
            return await _welcomeMessage.IsVisibleAsync(new() { Timeout = 5000 });
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Get the main heading text
    /// </summary>
    public async Task<string> GetHeadingTextAsync()
    {
        return await _pageHeading.TextContentAsync() ?? string.Empty;
    }

    /// <summary>
    /// Check if recent articles section is visible
    /// </summary>
    public async Task<bool> HasRecentArticlesAsync()
    {
        try
        {
            return await _recentArticles.IsVisibleAsync(new() { Timeout = 5000 });
        }
        catch
        {
            return false;
        }
    }
}